### PR TITLE
HPCC-21765 Allow global client dafilesrv defaults to be configured

### PR DESCRIPTION
**Pull request by jakesmith**
Allow: dafsConnectTimeoutSeconds, dafsConnectRetries,
dafsMaxReceiveTimeSeconds and dafsConnectFailRetrySeconds to be
configured via environment.conf

Also reduce default connect fail retry time to 10 seconds,
previously it was 15 minutes.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: fs/dafsclient/rmtclient.cpp [hunk 1]
```diff
@@ -79,8 +79,26 @@ void cleanupDaFsSocket(ISocket *sock)
 ///////////////////////////
 
 
-static unsigned maxConnectTime = 0;
-static unsigned maxReceiveTime = 0;
+static const unsigned defaultDafsConnectTimeoutSeconds=100;
```
Comment (jakesmith) [inline fs/dafsclient/rmtclient.cpp]: NB: this was previously default to 0 (maxConnectTime), which caused connectSocket to use ISocket::connect without a timeout, however, ISocket::connect itself has a default timeout of 100 seconds.
Hence I have defaulted the dafilesrv connect timeout to 100 here.

File: fs/dafsclient/rmtclient.cpp [hunk 2]
```diff
@@ -79,8 +79,26 @@ void cleanupDaFsSocket(ISocket *sock)
...
+static const unsigned defaultDafsConnectRetries=2;
+static const unsigned defaultDafsMaxRecieveTimeSeconds=0;
+static const unsigned defaultDafsConnectFailRetrySeconds=10;
```
Comment (jakesmith) [inline fs/dafsclient/rmtclient.cpp]: As per JIRA/PR comment, this reduces the default down to 10 seconds, previously it was 15 minutes (DAFS_CONNECT_FAIL_RETRY_TIME)

File: fs/dafsclient/rmtclient.cpp [hunk 3]
```diff
@@ -610,499 +627,492 @@ void clientAddSocketToCache(SocketEndpoint &ep,ISocket *socket)
...
-				if (tm.timemon->timedout(&remaining))
-					throwJSocketException(JSOCKERR_connection_failed);
-				socket.setown(ISocket::connect_timeout(ep,remaining));
-			}
-			else
-				socket.setown(ISocket::connect(ep));
-			if (ep.port == securitySettings.daFileSrvSSLPort)
-			{
+    if (!connectTimeoutMs)
+        connectTimeoutMs = dafsConnectTimeoutMs;
```
Comment (jakesmith) [inline fs/dafsclient/rmtclient.cpp]: NB: this is a slight (deliberate) change in semantics.
Previously if connectTimeoutMs was supplied and higher than the global default, it was effectively ignored.
After this change a supplied connectTimeout overrides the global default.

NB2: The only place a timeout was supplied it was (5000) it was always less than the hard-wired default, so this change does not effect any existing code.

File: fs/dafsclient/rmtclient_impl.hpp [hunk 4]
```diff
@@ -155,7 +155,7 @@ class CRemoteBase : public CSimpleInterfaceOf<IDaFsConnection>
     static unsigned         lastfailtime;
     DAFSConnectCfg          connectMethod;
 
-    void connectSocket(SocketEndpoint &ep, unsigned localConnectTime=0, unsigned localRetries=0);
+    void connectSocket(SocketEndpoint &ep, unsigned connectTimeoutMs=0, unsigned connectRetries=INFINITE);
```
Comment (jakesmith) [inline fs/dafsclient/rmtclient_impl.hpp]: NB: changed default to max retries value, as 0 is a legitimate override value for retries.

**Comment by jakesmith** [comment]
@mckellyln - please review

File: initfiles/etc/DIR_NAME/environment.conf.in [hunk 5]
```diff
@@ -66,3 +66,13 @@ additionalPlugins=python2
...
+
+# Dafilesrv: default client side connection settings (NB: 0 = disable/use system defaults)
+#dafsConnectTimeoutSeconds=100
+#dafsConnectRetries=2
+#dafsMaxReceiveTimeSeconds=0
+
+# Dafilesrv: set to change number of seconds before retrying an unresponsive dafilesrv connection (default 10 seconds)
+# NB: for now this only applies to the last cached server 
+#dafsConnectFailRetrySeconds=10
+
```
Comment (mckellyln) [inline initfiles/etc/DIR_NAME/environment.conf.in]: If you add to/change the environment.conf file you should also update:
initfiles/etc/DIR_NAME/environment.xml.in (in community and LN branches)
testing/regress/environment.xml.in
docs/*/HPCCSystemAdmin/HPCCSystemAdministratorsGuide.xml
initfiles/componentfiles/configschema/xsd/envsettings.xsd

File: fs/dafsclient/rmtclient.cpp [hunk 6]
```diff
@@ -79,8 +79,26 @@ void cleanupDaFsSocket(ISocket *sock)
...
+static const unsigned defaultDafsConnectFailRetrySeconds=10;
+
+static unsigned dafsConnectTimeoutMs = defaultDafsConnectTimeoutSeconds * 1000;
+static unsigned dafsConnectRetries = defaultDafsConnectRetries;
+static unsigned dafsMaxReceiveTimeMs = defaultDafsMaxRecieveTimeSeconds * 1000;
+static unsigned dafsConnectFailRetryTimeMs = defaultDafsConnectFailRetrySeconds * 1000;
+
+MODULE_INIT(INIT_PRIORITY_DAFSCLIENT)
+{
+    const IProperties &confProps = queryEnvironmentConf();
```
Comment (mckellyln) [inline fs/dafsclient/rmtclient.cpp]: Is it ok to call queryEnvironmentConf() this early ?

File: fs/dafsclient/rmtclient.cpp (hunk 3 above)
Comment (mckellyln) [reply fs/dafsclient/rmtclient.cpp]: Is there a minimum on the connect time ?  Perhaps we shoud make sure it is at least some minimum ?

**Comment by mckellyln** [comment]
@jakesmith looks good but have a few questions.
It needs a rebase/manual fixup after the recent update of rmtclient.cpp.
(Sorry, if we merged this one first it might have been easier)

File: fs/dafsclient/rmtclient.cpp (hunk 3 above)
Comment (jakesmith) [reply fs/dafsclient/rmtclient.cpp]: Not sure about that. If caller wants a very low minimum that's up to them.

File: initfiles/etc/DIR_NAME/environment.conf.in (hunk 5 above)
Comment (jakesmith) [reply initfiles/etc/DIR_NAME/environment.conf.in]: Agreed, but environment.xml not changed so esting/regress/environment.xml.in and LN changes not required.

Will update HPCCSystemAdministratorsGuide.xml and envsettings.xsd

**Comment by jakesmith** [comment]
@mckellyln - rebased, replied to comments.

File: initfiles/etc/DIR_NAME/environment.conf.in (hunk 5 above)
Comment (mckellyln) [reply initfiles/etc/DIR_NAME/environment.conf.in]: The top of the xml matches the conf - so I think its good to have them continue to match.

File: fs/dafsclient/rmtclient.cpp (hunk 3 above)
Comment (mckellyln) [reply fs/dafsclient/rmtclient.cpp]: connect_timeout() uses timeout passed in, connect_wait() ups timeout to >= 10 sec.

**Comment by mckellyln** [comment]
@jakesmith looks good.
I think the xml should be updated as well to match as then it remains consistent.  But I agree it is not mandatory since settings are read from .conf file.

**Review by mckellyln — APPROVED** [review]
@jakesmith approved - just my personal opinion that the xml files should be updated to match .conf.

File: initfiles/etc/DIR_NAME/environment.conf.in (hunk 5 above)
Comment (jakesmith) [reply initfiles/etc/DIR_NAME/environment.conf.in]: hm, yes agreed, though not sure anything uses them.
Nevertheless I'll update.

Comment (jakesmith) [reply initfiles/etc/DIR_NAME/environment.conf.in]: hmm there's loads of settings that haven't been put in there, making me doubt it's usefulness all the more.

Comment (jakesmith) [reply initfiles/etc/DIR_NAME/environment.conf.in]: Given that only some duplicated into <EnvSettings> that are in environment.conf, and there a bunch of others which are solely in <EnvSettings> (not in environment.conf), I won't add more duplicates into environment.xml

I think probably want needs to happen, is that the duplicates, that afaics are never read fromenvironment.xml, needs clearing up and removing.
Needs another JIRA to look at that though.

Comment (mckellyln) [reply initfiles/etc/DIR_NAME/environment.conf.in]: ok, sure.

**Comment by mckellyln** [comment]
@jakesmith approved

**Comment by jakesmith** [comment]
I've opened a separate JIRA to look at the fact that environment.xml <EnvSettings> has any [unused] duplicate settings from environment.conf - if they are never used, they should be deleted: https://track.hpccsystems.com/browse/HPCC-22197

**Comment by jakesmith** [comment]
@richardkchapman - please merge.

**Pull request merged**